415 on issueBulkOperations.getAvailableTransitions /rest/api/3/bulk/issues/transition

This feels like a bug on jira's side where that endpoint returns a status 415 Unsupported Media type error.
If I add header `Content-Type: application/json` then it works.
I find this a bit weird because this is a GET endpoint it should not required a Content-Type since there's no body
https://stackoverflow.com/a/16693884



```
Caused by: AxiosError: Request failed with status code 415
 ❯ settle node_modules/.store/axios-npm-1.8.4-8cf735eb2b/package/lib/core/settle.js:19:12
 ❯ IncomingMessage.handleStreamEnd node_modules/.store/axios-npm-1.8.4-8cf735eb2b/package/lib/adapters/http.js:599:11
 ❯ Axios.request node_modules/.store/axios-npm-1.8.4-8cf735eb2b/package/lib/core/Axios.js:45:41

⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
Serialized Error: {
  description: undefined,
  number: undefined,
  fileName: undefined,
  lineNumber: undefined,
  columnNumber: undefined,
  config: {
    transitional: {
      silentJSONParsing: true,
      forcedJSONParsing: true,
      clarifyTimeoutError: false
    },
    adapter: ['xhr', 'http', 'fetch'],
    transformRequest: ['Function<transformRequest>'],
    transformResponse: ['Function<transformResponse>'],
    timeout: +0,
    xsrfCookieName: 'XSRF-TOKEN',
    xsrfHeaderName: 'X-XSRF-TOKEN',
    maxContentLength: -1,
    maxBodyLength: -1,
    env: {
      FormData: 'Function<FormData>',
      Blob: 'Function<Blob>'
    },
    validateStatus: 'Function<validateStatus>',
    headers: {
      Accept: 'application/json, text/plain, */*',
      Authorization: 'Basic ***',
      'User-Agent': 'axios/1.8.4',
      'Accept-Encoding': 'gzip, compress, deflate, br'
    },
    paramsSerializer: {
      serialize: 'Function<bound paramSerializer>'
    },
    baseURL: 'https://maintainx.atlassian.net',
    url: '/rest/api/3/bulk/issues/transition',
    method: 'get',
    params: {
      issueIdsOrKeys: ['JT-36', 'JT-37']
    },
    allowAbsoluteUrls: true
  },
  code: 'ERR_BAD_REQUEST',
  status: 415
}
```

--- a/package.json
+++ b/package.json
@@ -19,7 +19,9 @@
     "library"
   ],
   "scripts": {
-    "build": "tsc",
+    "build": "npm run build:src && npm run build:tests",
+    "build:src": "tsc --project tsconfig.json",
+    "build:tests": "tsc --project tests/tsconfig.json",
     "prettier": "prettier --write src",
     "lint": "npm run lint:tests && npm run lint:examples && npm run lint:src:agile && npm run lint:src:clients && npm run lint:src:services && npm run lint:src:version2 && npm run lint:src:version3 && npm run lint:src:files",
     "lint:tests": "npm run lint:base -- tests",

--- a/src/version3/issueBulkOperations.ts
+++ b/src/version3/issueBulkOperations.ts
@@ -355,6 +355,9 @@ export class IssueBulkOperations {
         endingBefore: parameters.endingBefore,
         startingAfter: parameters.startingAfter,
       },
+      headers: {
+        'Content-Type': 'application/json',
+      },
     };
 
     return this.client.sendRequest(config, callback);

--- a/tests/integration/index.ts
+++ b/tests/integration/index.ts
@@ -1,2 +1,2 @@
-export * from '@tests/constants';
-export * as Utils from '@tests/utils';
+export * from '@tests/integration/constants';
+export * as Utils from '@tests/integration/utils';

--- a/tests/integration/utils/createSoftwareProject.ts
+++ b/tests/integration/utils/createSoftwareProject.ts
@@ -6,7 +6,13 @@ export const createSoftwareProject = async () => {
   const client = getVersion2Client();
   const currentUser = await client.myself.getCurrentUser();
 
-  if (!currentUser.accountId) throw new Error("Couldn't get the current user's account ID", { cause: { currentUser } });
+  if (!currentUser.accountId) {
+    throw new Error(
+      "Couldn't get the current user's account ID",
+      // @ts-expect-error -- Requires to use lib ES2022+ in tsconfig to use "cause"
+      { cause: { currentUser } },
+    );
+  }
 
   return client.projects
     .createProject({

--- a/tests/integration/version2/issueComments.test.ts
+++ b/tests/integration/version2/issueComments.test.ts
@@ -43,7 +43,7 @@ test.sequential('should update comment', async ({ expect }) => {
 
   const updatedComment = await client.issueComments.updateComment({
     issueIdOrKey: issue.key,
-    id: comment.id,
+    id: comment.id!,
     comment: 'updated comment',
   });
 

--- a/tests/integration/version3/issueBulkOperations.test.ts
+++ b/tests/integration/version3/issueBulkOperations.test.ts
@@ -1,0 +1,75 @@
+import { afterAll, beforeAll, test } from 'vitest';
+import { Constants } from '@tests/integration/constants';
+import { cleanupEnvironment, getVersion3Client, prepareEnvironment } from '@tests/integration/utils';
+import { CreatedIssue } from '@jirajs/version3/models';
+
+const client = getVersion3Client();
+let createdIssues: CreatedIssue[] = [];
+
+beforeAll(async () => {
+  await prepareEnvironment();
+});
+
+afterAll(async () => {
+  await cleanupEnvironment();
+});
+
+test.sequential('should create test issues', async ({ expect }) => {
+  function createIssue(num: number) {
+    return client.issues.createIssue({
+      fields: {
+        summary: `${Constants.testIssueSummary} ${num}`,
+        issuetype: { name: 'Task' },
+        project: { key: Constants.testProjectKey },
+        description: {
+          type: 'doc',
+          version: 1,
+          content: [
+            {
+              type: 'paragraph',
+              content: [
+                {
+                  text: `${Constants.testIssueDescription} ${num}`,
+                  type: 'text',
+                },
+              ],
+            },
+          ],
+        },
+      },
+    });
+  }
+
+  createdIssues = await Promise.all([createIssue(1), createIssue(2)]);
+
+  expect(createdIssues).toHaveLength(2);
+  expect(createdIssues[0].key).toContain(Constants.testProjectKey);
+  expect(createdIssues[1].key).toContain(Constants.testProjectKey);
+});
+
+test.sequential('should get available transitions for multiple issues', async ({ expect }) => {
+  const result = await client.issueBulkOperations.getAvailableTransitions({
+    issueIdsOrKeys: createdIssues.map(issue => issue.key),
+  });
+
+  expect(result).toBeDefined();
+  expect(result.availableTransitions).toBeDefined();
+  expect(Array.isArray(result.availableTransitions)).toBeTruthy();
+
+  if (result.availableTransitions && result.availableTransitions.length > 0) {
+    const firstWorkflow = result.availableTransitions[0];
+
+    expect(Array.isArray(firstWorkflow.transitions)).toBeTruthy();
+    expect(Array.isArray(firstWorkflow.issues)).toBeTruthy();
+    expect(typeof firstWorkflow.isTransitionsFiltered).toBe('boolean');
+  }
+});
+
+test.sequential('should cleanup test issues', async ({ expect }) => {
+  const result = await client.issueBulkOperations.submitBulkDelete({
+    selectedIssueIdsOrKeys: createdIssues.map(issue => issue.key),
+  });
+
+  expect(result).toBeDefined();
+  expect(result.taskId).toBeDefined();
+});

--- a/tests/tsconfig.json
+++ b/tests/tsconfig.json
@@ -1,0 +1,15 @@
+{
+  "extends": "../tsconfig.json",
+  "compilerOptions": {
+    "baseUrl": ".",
+    "outDir": "./out",
+    "noEmit": true,
+    "paths": {
+      "@jirajs": ["../src"],
+      "@jirajs/*": ["../src/*"],
+      "@tests": ["."],
+      "@tests/*": ["*"]
+    }
+  },
+  "include": ["integration", "unit"]
+}

--- a/tests/unit/version2/issueSearch.test.ts
+++ b/tests/unit/version2/issueSearch.test.ts
@@ -75,7 +75,9 @@ test('searchForIssuesUsingJqlEnhancedSearch should calls without parameters', ({
   const client = new Version2Client(config);
   const sendRequestStub = sinon.stub(client, 'sendRequest');
 
-  client.issueSearch.searchForIssuesUsingJqlEnhancedSearch({});
+  client.issueSearch.searchForIssuesUsingJqlEnhancedSearch({
+    jql: '',
+  });
 
   expect(sendRequestStub.calledOnce).toBeTruthy();
 });

--- a/tests/unit/version3/issueBulkOperations.test.ts
+++ b/tests/unit/version3/issueBulkOperations.test.ts
@@ -1,0 +1,33 @@
+import * as sinon from 'sinon';
+import { test } from 'vitest';
+import { Version3Client } from '@jirajs';
+
+const config = {
+  host: 'http://localhost',
+};
+
+test('getAvailableTransitions should accept follow parameters', ({ expect }) => {
+  const client = new Version3Client(config);
+  const sendRequestStub = sinon.stub(client, 'sendRequest');
+
+  client.issueBulkOperations.getAvailableTransitions({
+    issueIdsOrKeys: ['PROJ-1', 'PROJ-2'],
+    startingAfter: 'cursor1',
+    endingBefore: 'cursor2',
+  });
+
+  expect(sendRequestStub.calledOnce).toBeTruthy();
+
+  const callArgument = sendRequestStub.getCall(0).args[0];
+
+  expect(callArgument.url).toBe('/rest/api/3/bulk/issues/transition');
+  expect(callArgument.method).toBe('GET');
+  expect(callArgument.params).toStrictEqual({
+    issueIdsOrKeys: ['PROJ-1', 'PROJ-2'],
+    startingAfter: 'cursor1',
+    endingBefore: 'cursor2',
+  });
+  expect(callArgument.headers).toStrictEqual({
+    'Content-Type': 'application/json',
+  });
+});

--- a/tests/unit/version3/issueSearch.test.ts
+++ b/tests/unit/version3/issueSearch.test.ts
@@ -75,7 +75,9 @@ test('searchForIssuesUsingJqlEnhancedSearch should calls without parameters', ({
   const client = new Version3Client(config);
   const sendRequestStub = sinon.stub(client, 'sendRequest');
 
-  client.issueSearch.searchForIssuesUsingJqlEnhancedSearch({});
+  client.issueSearch.searchForIssuesUsingJqlEnhancedSearch({
+    jql: '',
+  });
 
   expect(sendRequestStub.calledOnce).toBeTruthy();
 });

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -9,12 +9,6 @@
       "DOM"
     ],
     "baseUrl": ".",
-    "paths": {
-      "@jirajs": ["src"],
-      "@jirajs/*": ["src/*"],
-      "@tests": ["tests"],
-      "@tests/*": ["tests/*"]
-    },
     "declaration": true,
     "importHelpers": true,
     "strict": true,

